File: frontend/src/utils/signals.js
/**
 * Signals Utility - Fetch symptom and context signals from Firestore
 */

const API_BASE = 'http://localhost:5000/api'

export async function getSignals() {
  try {
    const response = await fetch(`${API_BASE}/signals/`)
    if (!response.ok) throw new Error('Failed to fetch signals')
    const data = await response.json()
    return data.data || {}
  } catch (error) {
    console.error('Error fetching signals:', error)
    return { symptomSignals: {}, contextSignals: {} }
  }
}

export async function getSymptomCategory(category) {
  try {
    const response = await fetch(`${API_BASE}/signals/symptom/${category}`)
    if (!response.ok) throw new Error(`Failed to fetch ${category} symptoms`)
    const data = await response.json()
    return data.symptoms || []
  } catch (error) {
    console.error(`Error fetching ${category} symptoms:`, error)
    return []
  }
}

export async function getContextCategory(category) {
  try {
    const response = await fetch(`${API_BASE}/signals/context/${category}`)
    if (!response.ok) throw new Error(`Failed to fetch ${category} contexts`)
    const data = await response.json()
    return data.contexts || []
  } catch (error) {
    console.error(`Error fetching ${category} contexts:`, error)
    return []
  }
}

export async function initializeSignals() {
  try {
    const response = await fetch(`${API_BASE}/signals/init`, { method: 'POST' })
    if (!response.ok) throw new Error('Failed to initialize signals')
    const data = await response.json()
    return data
  } catch (error) {
    console.error('Error initializing signals:', error)
    return null
  }
}

export async function updateSignals(symptomSignals, contextSignals) {
  try {
    const response = await fetch(`${API_BASE}/signals/`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ symptomSignals, contextSignals })
    })
    if (!response.ok) throw new Error('Failed to update signals')
    const data = await response.json()
    return data
  } catch (error) {
    console.error('Error updating signals:', error)
    return null
  }
}

// English + Hindi + Telugu emergency signals (extensible)
export const SYMPTOM_SIGNALS = {
  cardiac: [
    { text: "chest pain", lang: "en" },
    { text: "crushing chest", lang: "en" },
    { text: "pain in left arm", lang: "en" },
    { text: "jaw pain", lang: "en" },
    { text: "heart attack", lang: "en" },
    { text: "tightness in chest", lang: "en" },

    { text: "सीने में दर्द", lang: "hi" },
    { text: "दिल का दौरा", lang: "hi" },
    { text: "बाएं हाथ में दर्द", lang: "hi" },
    { text: "छाती में जकड़न", lang: "hi" },

    { text: "ఛాతిలో నొప్పి", lang: "te" },
    { text: "గుండె నొప్పి", lang: "te" },
    { text: "హృదయాఘాతం", lang: "te" },
  ],

  neurological: [
    "slurred speech",
    "stroke",
    "seizure",
    "arm weakness",

    "बोलने में कठिनाई",
    "लकवा",
    "दौरा",

    "మాట తడబడటం",
    "పక్షవాతం",
    "ఫిట్స్",
  ],

  respiratory: [
    { text: "shortness of breath", lang: "en" },
    { text: "cannot breathe", lang: "en" },
    { text: "gasping for air", lang: "en" },

    { text: "सांस लेने में कठिनाई", lang: "hi" },
    { text: "सांस नहीं आ रही", lang: "hi" },

    { text: "శ్వాస తీసుకోవడంలో ఇబ్బంది", lang: "te" },
    { text: "ఊపిరి రావడం లేదు", lang: "te" },
  ],

  bleeding: [
    "heavy bleeding",
    "vomiting blood",

    "बहुत ज्यादा खून बहना",
    "खून की उल्टी",

    "అధిక రక్తస్రావం",
    "రక్త వాంతులు",
  ],

  trauma: [
    "car accident",
    "head injury",

    "कार दुर्घटना",
    "सिर में चोट",

    "రోడ్డు ప్రమాదం",
    "తల గాయం",
  ],
};

export const CONTEXT_SIGNALS = {
  sudden: ["sudden", "अचानक", "అకస్మాత్తుగా"],
  worsening: ["getting worse", "बढ़ रहा है", "మరింత ఎక్కువవుతోంది"],
  exertion: ["walking", "exercise", "चलते समय", "నడుస్తున్నప్పుడు"],
  duration: ["for hours", "कई घंटों से", "గంటలుగా"],
};

File: backend/routes/signals.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Dict, List

try:
    from firebase_admin import firestore  # type: ignore
    db = firestore.client()
except Exception:
    db = None

router = APIRouter()

class SignalDefinition(BaseModel):
    symptomSignals: Dict[str, List[str]]
    contextSignals: Dict[str, List[str]]

# Default signal definitions
DEFAULT_SIGNALS = {
    "symptomSignals": {
        "cardiac": [
            "chest pain",
            "crushing chest",
            "pain in left arm",
            "jaw pain",
            "heart attack",
            "tightness in chest"
        ],
        "neurological": [
            "slurred speech",
            "facial droop",
            "numbness on one side",
            "arm weakness",
            "stroke",
            "seizure",
            "loss of vision",
            "confused",
            "worst headache of my life"
        ],
        "respiratory": [
            "cannot breathe",
            "shortness of breath",
            "gasping for air",
            "turning blue",
            "choking"
        ],
        "bleeding": [
            "heavy bleeding",
            "vomiting blood",
            "coughing up blood",
            "deep wound",
            "stab wound",
            "gunshot"
        ],
        "trauma": [
            "fall from height",
            "car accident",
            "head injury",
            "broken bone"
        ]
    },
    "contextSignals": {
        "sudden": ["sudden", "abrupt", "instantly"],
        "worsening": ["getting worse", "rapidly", "unbearable"],
        "exertion": ["while running", "during exercise", "walking up stairs"],
        "duration": ["for hours", "persisting", "constant"]
    }
}

@router.post("/init")
async def initialize_signals():
    """Initialize signal definitions in Firestore."""
    if not db:
        raise HTTPException(status_code=503, detail="Database not available")
    
    try:
        db.collection("config").document("signals").set(DEFAULT_SIGNALS)
        return {
            "success": True,
            "message": "Signal definitions initialized in Firestore",
            "data": DEFAULT_SIGNALS
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to initialize signals: {str(e)}")

@router.get("/")
async def get_signals():
    """Retrieve signal definitions from Firestore."""
    if not db:
        raise HTTPException(status_code=503, detail="Database not available")
    
    try:
        doc = db.collection("config").document("signals").get()
        if doc.exists:
            return {
                "success": True,
                "data": doc.to_dict()
            }
        else:
            # Return default if not found
            return {
                "success": True,
                "data": DEFAULT_SIGNALS
            }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve signals: {str(e)}")

@router.put("/")
async def update_signals(signals: SignalDefinition):
    """Update signal definitions in Firestore."""
    if not db:
        raise HTTPException(status_code=503, detail="Database not available")
    
    try:
        data = signals.dict()
        db.collection("config").document("signals").set(data)
        return {
            "success": True,
            "message": "Signal definitions updated",
            "data": data
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update signals: {str(e)}")

@router.get("/symptom/{category}")
async def get_symptom_category(category: str):
    """Get symptoms for a specific category."""
    if not db:
        raise HTTPException(status_code=503, detail="Database not available")
    
    try:
        doc = db.collection("config").document("signals").get()
        if doc.exists:
            data = doc.to_dict()
            symptoms = data.get("symptomSignals", {}).get(category, [])
            if symptoms:
                return {
                    "success": True,
                    "category": category,
                    "symptoms": symptoms
                }
        
        # Fallback to default
        symptoms = DEFAULT_SIGNALS["symptomSignals"].get(category, [])
        return {
            "success": True,
            "category": category,
            "symptoms": symptoms
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve symptoms: {str(e)}")

@router.get("/context/{category}")
async def get_context_category(category: str):
    """Get context signals for a specific category."""
    if not db:
        raise HTTPException(status_code=503, detail="Database not available")
    
    try:
        doc = db.collection("config").document("signals").get()
        if doc.exists:
            data = doc.to_dict()
            contexts = data.get("contextSignals", {}).get(category, [])
            if contexts:
                return {
                    "success": True,
                    "category": category,
                    "contexts": contexts
                }
        
        # Fallback to default
        contexts = DEFAULT_SIGNALS["contextSignals"].get(category, [])
        return {
            "success": True,
            "category": category,
            "contexts": contexts
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve contexts: {str(e)}")
